runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/SingletonDockerContainerManager.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.environment;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.util.concurrent.SettableFuture;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClientControlService;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.runners.fnexecution.provisioning.StaticGrpcProvisionService;
+
+/**
+ * A {@link ContainerManager} which manages a single Docker Container.
+ *
+ * <p>The container is initialized on the first call to {@link #getEnvironment(Environment)}. Future
+ * calls will fail if there is more than one environment.
+ */
+public class SingletonDockerContainerManager implements ContainerManager {
+  public static ContainerManager forServers(
+      GrpcFnServer<SdkHarnessClientControlService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer) {
+    return new SingletonDockerContainerManager(
+        controlServiceServer,
+        loggingServiceServer,
+        retrievalServiceServer,
+        provisioningServiceServer);
+  }
+
+  private final SettableFuture<DockerContainerEnvironment> process = SettableFuture.create();
+
+  private final GrpcFnServer<SdkHarnessClientControlService> controlServiceServer;
+  private final GrpcFnServer<GrpcLoggingService> loggingServiceServer;
+  private final GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer;
+  private final GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer;
+
+  private SingletonDockerContainerManager(
+      GrpcFnServer<SdkHarnessClientControlService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer) {
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+  }
+
+  @Override
+  public RemoteEnvironment getEnvironment(Environment environment) throws Exception {
+    if (process.isDone()) {
```

Comment:
Is this intended to be used concurrently? If so, this will need to be protected. FYI, for the Flink runner, we will almost certainly want to use this concurrently.